### PR TITLE
 testharness: only run tests in CWD subtree

### LINKED ISSUE #10209: run_tests to use specific test file
## Rationale

When developing it would be useful to run the tests from a given tests spec.

## Description

Add a means to pass a "tests" file to the run_tests script.

## Impact

None

### PR DESCRIPTION
**Pull request by rwcarlsen**
Instead of the run_tests script running all tests in its dir subtree
using a specific named binary, modify the test harness to just run tests
in the current working directory's subtree.  The binary is discovered
walking of the parent directory hierarchy looking for a "testroot" file.
This file contains the name of the binary to be used.  "run_tests"
scripts now can all be identical with no app-specific content.

Also provide a "moosetest" script that users can add to their path that
will always run tests in the CWD for *any* moose app/binary - it just
hunts for a "testroot" file.

fixes #10209

### CODE REVIEW

**Comment by rwcarlsen** [comment]
This should be ready for a final look/merge.

File: scripts/moosetest [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python
+import sys, os
+
+MOOSE_DIR = os.environ.get('MOOSE_DIR', os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
+sys.path.append(os.path.join(MOOSE_DIR, 'python'))
+os.environ['MOOSE_DIR'] = MOOSE_DIR
+os.environ['PYTHONPATH'] = os.environ.get('PYTHONPATH', '') + ':' + os.path.join(MOOSE_DIR, 'python')
```
Comment (brianmoose) [inline scripts/moosetest]: Do these 2 need to be set here? They are set in the TestHarness right?

File: python/TestHarness/TestHarness.py [hunk 2]
```diff
@@ -4,33 +4,66 @@
...
-            harness = TestHarness(argv, app_name, moose_dir)
+            harness = TestHarness(argv, moose_dir, app_name=app_name, wd=wd)
 
         harness.findAndRunTests()
-
         sys.exit(harness.error_code)
 
+    def __init__(self, argv, moose_dir, app_name=None, wd=None):
+        os.environ['MOOSE_DIR'] = moose_dir
+        os.environ['PYTHONPATH'] = os.environ.get('PYTHONPATH', '') + ':' + os.path.join(moose_dir, 'python')
```
Comment (brianmoose) [inline python/TestHarness/TestHarness.py]: Maybe prepend to PYTHONPATH?

**Review by brianmoose — CHANGES_REQUESTED** [review]

File: scripts/moosetest (hunk 1 above)
Comment (rwcarlsen) [reply scripts/moosetest]: fixed

**Comment by permcody** [comment]
I tested this in the modules and it didn't work quite like I expected. I built in the modules directory but then when into the phase_field directory to run tests. I actually ran `./run_tests` and I expected to get the error from the test harness saying that I didn't have an executable built. That's not what happened, it actually ran the tests in the phase_field directory but used the combined module to do it. 

Is that wrong? It's debatable, but I'm going to say that it's wrong. Here's why: If I'm actually building a pure phase_field input, I'm going to want to run it with the phase_field executable to make sure that I don't get any cross module pollination. However, there was no indication this was happening, it was falling back to the combined module. In other words, it blew past the test root and found the one further up stream. We need to design for this.

File: modules/navier_stokes/run_tests [hunk 3]
```diff
@@ -1,23 +1,8 @@
...
-#### Set the name of the application here and moose directory relative to the application
-app_name = 'navier_stokes'
-
-MODULE_DIR = os.path.abspath('..')
-MOOSE_DIR = os.path.abspath(os.path.join(MODULE_DIR, '..'))
-#### See if MOOSE_DIR is already in the environment instead
-if "MOOSE_DIR" in os.environ:
-  MOOSE_DIR = os.environ['MOOSE_DIR']
-
+MOOSE_DIR = os.path.abspath(os.environ.get('MOOSE_DIR', os.path.join('..', '..')))
```
Comment (brianmoose) [inline modules/navier_stokes/run_tests]: Needs to be updated

File: python/run_tests [hunk 4]
```diff
@@ -1,20 +1,8 @@
...
-#### Set the name of the application here and moose directory relative to the application
-app_name = 'moose_python'
-
-MOOSE_DIR = os.path.abspath(os.path.join('..'))
-#### See if MOOSE_DIR is already in the environment instead
-if "MOOSE_DIR" in os.environ:
-    MOOSE_DIR = os.environ['MOOSE_DIR']
+import sys, os
 
+MOOSE_DIR = os.environ.get('MOOSE_DIR', '..')
```
Comment (brianmoose) [inline python/run_tests]: absolute path

**Review by brianmoose — CHANGES_REQUESTED** [review]
Since all these run_tests are the same now, would it make more sense to just symbolic link to the scripts/run_tests file?

**Review by brianmoose — APPROVED** [review]

**Comment by rwcarlsen** [comment]
Thanks for the thorough review @brianmoose.  Your sharp eye enables my lazy coding.

**Comment by bwspenc** [comment]
It must have been a while since I've done real work, because I just noticed that there's this annoying new 'testroot' file sitting in the same directory as the 'test' directory that messes with my tab completion. When I did 'git blame', I was shocked to find that @rwcarlsen created that file.  Weren't you the one who was a staunch advocate of having good tab completion with the whole test/tests change? Now I can't just do 'cd te <tab> <tab>' to get into the 'test/tests' directory.  Any way we could pick another filename?  Anything that doesn't start with 'test' is fine with me.  Maybe '.testroot'?

**Comment by permcody** [comment]
😂

**Pull request merged**